### ISSUE TITLE
could you please report your training cost?

### ISSUE THREAD

**Issue by lucywang720** [allenai/open-instruct#72]
seems that it takes a long time to train tulu, I am now trying to train sulu based on llama-2-7b, but I have to use 8*A100 40G, and I also need to modify --mixed_precision bf16 to --mixed_precision fp16 , but it output like that:

  0%|          | 2/12525 [03:18<344:16:10, 98.97s/it] 10/13/2023 06:31:56 - INFO - __main__ -   Step: 2, LR: 0.0, Loss: 1.25099515914917

is that 344h?why does it take so long?

**Comment by lucywang720** [comment]
my script:

export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7

MODEL_SIZE=7B
NUM_GPUS=8
BATCH_SIZE_PER_GPU=1
TOTAL_BATCH_SIZE=128
GRADIENT_ACC_STEPS=$(($TOTAL_BATCH_SIZE/$NUM_GPUS/$BATCH_SIZE_PER_GPU))
echo "Training llama model ${MODEL_SIZE} using $NUM_GPUS GPUs, $BATCH_SIZE_PER_GPU batch size per GPU, $GRADIENT_ACC_STEPS gradient accumulation steps"

accelerate launch \
    --mixed_precision fp16 \
    --num_machines 1 \
    --num_processes $NUM_GPUS \
    --use_deepspeed \
    --deepspeed_config_file ds_configs/stage3_no_offloading_accelerate.conf \
    open_instruct/finetune.py \
    --model_name_or_path /cluster/data0/whq/aa_llama2-7b-hf \
    --use_flash_attn \
    --use_lora \
    --lora_rank 256 \
    --lora_alpha 256 \
    --lora_dropout 0.1 \
    --tokenizer_name /cluster/data0/whq/aa_llama2-7b-hf \
    --use_slow_tokenizer \
    --train_file /cluster/home2/whq/open_instruct/data/processed/tulu_v2/tulu_v2_data.jsonl \
    --max_seq_length 2048 \
    --preprocessing_num_workers 16 \
    --checkpointing_steps epoch \
    --per_device_train_batch_size $BATCH_SIZE_PER_GPU \
    --gradient_accumulation_steps $GRADIENT_ACC_STEPS \
    --learning_rate 1e-4 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0. \
    --num_train_epochs 5 \
    --output_dir output/tulu_lora/ \
    --with_tracking \
    --report_to tensorboard \
    --logging_steps 1 &&

python open_instruct/merge_lora.py \
    --base_model_name_or_path /net/nfs.cirrascale/allennlp/yizhongw/hf_llama_models/${MODEL_SIZE} \
    --lora_model_name_or_path output/tulu_lora \
    --output_dir output/lora_test_merged/

**Comment by lucywang720** [comment]
my script

`export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7

MODEL_SIZE=7B
NUM_GPUS=8
BATCH_SIZE_PER_GPU=1
TOTAL_BATCH_SIZE=128
GRADIENT_ACC_STEPS=$(($TOTAL_BATCH_SIZE/$NUM_GPUS/$BATCH_SIZE_PER_GPU))
echo "Training llama model ${MODEL_SIZE} using $NUM_GPUS GPUs, $BATCH_SIZE_PER_GPU batch size per GPU, $GRADIENT_ACC_STEPS gradient accumulation steps"

accelerate launch \
    --mixed_precision fp16 \
    --num_machines 1 \
    --num_processes $NUM_GPUS \
    --use_deepspeed \
    --deepspeed_config_file ds_configs/stage3_no_offloading_accelerate.conf \
    open_instruct/finetune.py \
    --model_name_or_path /cluster/data0/whq/aa_llama2-7b-hf \
    --use_flash_attn \
    --use_lora \
    --lora_rank 256 \
    --lora_alpha 256 \
    --lora_dropout 0.1 \
    --tokenizer_name /cluster/data0/whq/aa_llama2-7b-hf \
    --use_slow_tokenizer \
    --train_file /cluster/home2/whq/open_instruct/data/processed/tulu_v2/tulu_v2_data.jsonl \
    --max_seq_length 2048 \
    --preprocessing_num_workers 16 \
    --checkpointing_steps epoch \
    --per_device_train_batch_size $BATCH_SIZE_PER_GPU \
    --gradient_accumulation_steps $GRADIENT_ACC_STEPS \
    --learning_rate 1e-4 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0. \
    --num_train_epochs 5 \
    --output_dir output/tulu_lora/ \
    --with_tracking \
    --report_to tensorboard \
    --logging_steps 1 &&

python open_instruct/merge_lora.py \
    --base_model_name_or_path /net/nfs.cirrascale/allennlp/yizhongw/hf_llama_models/${MODEL_SIZE} \
    --lora_model_name_or_path output/tulu_lora \
    --output_dir output/lora_test_merged/`

**Comment by lucywang720** [comment]
my script

`export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7

MODEL_SIZE=7B
NUM_GPUS=8
BATCH_SIZE_PER_GPU=1
TOTAL_BATCH_SIZE=128
GRADIENT_ACC_STEPS=$(($TOTAL_BATCH_SIZE/$NUM_GPUS/$BATCH_SIZE_PER_GPU))
echo "Training llama model ${MODEL_SIZE} using $NUM_GPUS GPUs, $BATCH_SIZE_PER_GPU batch size per GPU, $GRADIENT_ACC_STEPS gradient accumulation steps"

accelerate launch \
    --mixed_precision fp16 \
    --num_machines 1 \
    --num_processes $NUM_GPUS \
    --use_deepspeed \
    --deepspeed_config_file ds_configs/stage3_no_offloading_accelerate.conf \
    open_instruct/finetune.py \
    --model_name_or_path /cluster/data0/llama2-7b-hf \
    --use_flash_attn \
    --use_lora \
    --lora_rank 256 \
    --lora_alpha 256 \
    --lora_dropout 0.1 \
    --tokenizer_name /cluster/data0/llama2-7b-hf \
    --use_slow_tokenizer \
    --train_file /cluster/home2/whq/open_instruct/data/processed/tulu_v2/tulu_v2_data.jsonl \
    --max_seq_length 2048 \
    --preprocessing_num_workers 16 \
    --checkpointing_steps epoch \
    --per_device_train_batch_size $BATCH_SIZE_PER_GPU \
    --gradient_accumulation_steps $GRADIENT_ACC_STEPS \
    --learning_rate 1e-4 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0. \
    --num_train_epochs 5 \
    --output_dir output/tulu_lora/ \
    --with_tracking \
    --report_to tensorboard \
    --logging_steps 1 &&

python open_instruct/merge_lora.py \
    --base_model_name_or_path /net/nfs.cirrascale/allennlp/yizhongw/hf_llama_models/${MODEL_SIZE} \
    --lora_model_name_or_path output/tulu_lora \
    --output_dir output/lora_test_merged/`

**Comment by hamishivi** [comment]
Hi, the tulu mixture is quite large, so training is expected to take a while. Lora training a v2 model took roughly 13 hours on 32 80GB A100 GPUs for us, with a much smaller lora rank (64), so it might just be that this is slow.

The things I would try are:
1. double check your v2 mixture file, it should have around ~300k samples in it. If you're trying to train on much bigger dataset then that will slow things down.
2. try using qlora on top of this, and see if you can reduce the number of gradient accumulation steps? You can just do this with the `use_qlora` argument to `finetune.py`. Generally, we found that qlora doesn't degrade performance compared to lora.
3. try reducing the lora rank. We used a rank of 64 and alpha of 16 (recommended by the qlora authors). A smaller rank should speed things up.
